Profiler/CProfilerCallback.cpp

```diff
@@ -1,488 +1,555 @@
-}
-
-void CProfilerCallback::writeFunctionInfosToLog() {
-	// Must be called from synchronized context
-	traceLog.writeInlinedFunctionInfosToLog(&inlinedMethods);
-	inlinedMethods.clear();
-
-	traceLog.writeJittedFunctionInfosToLog(&jittedMethods);
-	jittedMethods.clear();
-}
-
-HRESULT CProfilerCallback::getFunctionInfo(FunctionID functionId, FunctionInfo* info) {
-	ModuleID moduleId = 0;
-	HRESULT hr = profilerInfo->GetFunctionInfo2(functionId, 0,
-		NULL, &moduleId, &info->functionToken, 0, NULL, NULL);
-
-	if (SUCCEEDED(hr) && moduleId != 0) {
-		AssemblyID assemblyId;
-		hr = profilerInfo->GetModuleInfo(moduleId, NULL, NULL,
-			NULL, NULL, &assemblyId);
-		if (SUCCEEDED(hr)) {
-			info->assemblyNumber = assemblyMap[assemblyId];
-		}
-	}
-
-	return hr;
-}
-
-int CProfilerCallback::writeFileVersionInfo(LPCWSTR assemblyPath, char* buffer, size_t bufferSize) {
-	DWORD infoSize = GetFileVersionInfoSizeW(assemblyPath, NULL);
-	if (!infoSize) {
-		return 0;
-	}
-
-	BYTE* versionInfo = new BYTE[infoSize];
-	if (!GetFileVersionInfoW(assemblyPath, NULL, infoSize, versionInfo)) {
-		return 0;
-	}
-
-	VS_FIXEDFILEINFO* fileInfo = NULL;
-	UINT fileInfoLength = 0;
-	if (!VerQueryValueW(versionInfo, L"\\", (void**)&fileInfo, &fileInfoLength)) {
-		return 0;
-	}
-
-	int	 version[4];
-	version[0] = HIWORD(fileInfo->dwFileVersionMS);
-	version[1] = LOWORD(fileInfo->dwFileVersionMS);
-	version[2] = HIWORD(fileInfo->dwFileVersionLS);
-	version[3] = LOWORD(fileInfo->dwFileVersionLS);
-
-	int writtenChars = sprintf_s(buffer, bufferSize, " FileVersion:%i.%i.%i.%i",
-		version[0], version[1], version[2], version[3]);
-
-	version[0] = HIWORD(fileInfo->dwProductVersionMS);
-	version[1] = LOWORD(fileInfo->dwProductVersionMS);
-	version[2] = HIWORD(fileInfo->dwProductVersionLS);
-	version[3] = LOWORD(fileInfo->dwProductVersionLS);
-
-	writtenChars += sprintf_s(buffer + writtenChars, bufferSize - writtenChars, " ProductVersion:%i.%i.%i.%i",
-		version[0], version[1], version[2], version[3]);
-
-	delete versionInfo;
-	return writtenChars;
-}
+#include "CProfilerCallback.h"
+#include "version.h"
+#include "UploadDaemon.h"
+#include "utils/StringUtils.h"
+#include "utils/WindowsUtils.h"
+#include "utils/Debug.h"
+#include <fstream>
+#include <algorithm>
+#include <winuser.h>
+#include <iostream>
+#include <utils/MethodEnter.h>
+
+#pragma comment(lib, "version.lib")
+#pragma intrinsic(strcmp,labs,strcpy,_rotl,memcmp,strlen,_rotr,memcpy,_lrotl,_strset,memset,_lrotr,abs,strcat)
+
+/**
+ * Serializes access to the singleton profiler instance when trying to shut it down.
+ * This prevents race conditions that can result in deadlocks that freeze the profiled process.
+ */
+class ShutdownGuard {
+public:
+	ShutdownGuard() {
+		// we never delete this critical section. This is not needed as it's cleaned up on process
+		// death like any other memory
+		InitializeCriticalSection(&section);
+	}
+
+	virtual ~ShutdownGuard() {}
```

Comment:
see [cpp core guidlines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c80-use-default-if-you-have-to-be-explicit-about-using-the-default-semantics)
```suggestion
	~ShutdownGuard() = default;
```